crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -421,6 +396,21 @@ impl CheckpointStore {
             }
         );
 
+        // Ensure we have processed all transactions contained in this checkpoint.
+        if !self.all_checkpoint_transactions_executed(contents)? {
+            return Err(SuiError::from(
+                "Checkpoint contains unexecuted transactions.",
+            ));
+        }
+
+        // TODO: ISSUE #2039
+        // - check the new checkpoint contents do not include already checkpointed transactions.
+        // - check that the sequence is complete, ie no missing dependencies in the checkpoint or
+        //   previous checkpoint for all transactions included.
+        // - create a cannonical causal order that is deterministic accross authorities and store
+        //   contents as such a list instead of a set.
+        // Probably we need access to the effects to do the above.
```

Comment:
I an authority cannot move to the next checkpoint now until it has processed all previous transactions. So by induction on the sequence of checkpoints, if we are at a checkpoint number now, all previous transactions since genesis are executed.